File: rust/crates/atelier-app/src/commands/export_report.rs
//! `report <slug>`: render one registered report as CSV.

use atelier_bench::contracts::exporter_contract::ExporterContract;
use atelier_bench::exporters::csv_exporter::CsvExporter;

use crate::state::AppState;

/// Render `slug` as CSV, or explain that it is not registered.
pub fn run(state: &AppState, slug: &str) -> anyhow::Result<String> {
    let report =
        atelier_bench::report(slug).ok_or_else(|| anyhow::anyhow!("unknown report {slug}"))?;

    Ok(CsvExporter::new().export(&report.rows(&state.data)))
}

File: go/app/commands/export_report.go
// Package commands carries the console commands.
package commands

import (
	"fmt"

	"atelier.example/lane/app"
	"atelier.example/lane/atelier/exporters"
	"atelier.example/lane/atelier/reports"
	"atelier.example/yamlish.v2"
)

// ExportReport renders one report through the exporter registered under the
// given extension. Both lookups are STRING-KEYED: neither the report type nor
// the exporter type is named here.
func ExportReport(state *app.State, slug string, extension string) (string, error) {
	report, found := reports.BySlug(slug)
	if !found {
		return "", fmt.Errorf("unknown report %q", slug)
	}
	exporter, found := exporters.ByExtension(extension)
	if !found {
		return "", fmt.Errorf("unknown format %q", extension)
	}
	return exporter.Export(report.Rows(state.Data)), nil
}

// ExportManifest renders the run manifest with the in-tree yamlish module.
// The import path ends in "yamlish.v2" but the package identifier is
// "yamlish", so the call below does not match the path's last segment.
func ExportManifest(state *app.State) string {
	return yamlish.Marshal(map[string]string{
		"calculator": state.Container.InvoiceCalculator().Name(),
		"orders":     fmt.Sprint(len(state.Data.Orders)),
		"reports":    fmt.Sprint(reports.Count()),
		"revenue":    fmt.Sprint(state.Data.RevenueCents()),
		"version":    yamlish.Version,
	})
}
